### PR TITLE
Implement callbacks

### LINKED ISSUE #11: Add custom implementation for all ForEach and DoWith methods
These are key to properly testing the plugins' callback-storing.

### PR DESCRIPTION
**Pull request by madmaxoft**
Provides special implementation for most APIs taking a callback.

Fixes #11 .

### CODE REVIEW

File: APIImpl/PluginManager.lua [hunk 1]
```diff
@@ -29,17 +42,118 @@ end
...
+
+--- Implementation of the cPluginManager:ForEachPlugin function
+-- There are two separate API endpoints for the same implementation
+local function ForEachPlugin(a_Simulator, a_Self, a_Callback)
+	local res = a_Simulator:processCallbackRequest({
+		Function = a_Callback,
+		ParamValues = { a_Self:GetCurrentPlugin() },
+		Notes = "cPluginManager:ForEachPlugin() callback",
+	})
+	return (res == true) and false or true  -- If the callback signalized an abort, return false
```
Comment (NiLSPACE) [inline APIImpl/PluginManager.lua]: This always returns true for me:

```
> =(true == true) and false or true
true
> =(false == true) and false or true
true
```

File: APIImpl/PluginManager.lua [hunk 2]
```diff
@@ -29,17 +42,118 @@ end
...
+
 return
 {
-	["<static> cPluginManager:AddHook(cPluginManager#PluginHook, function)"] = function (a_Simulator, a_Self, a_HookType, a_Callback)
+	["cPlugin:GetLocalFolder()"] = function(a_Simulator, a_Self)
+		return getmetatable(a_Self).simulatorInternal_pluginPath or a_Simulator:createInstance({Type = "string"})
+	end,
+
+	["cPluginLua:AddWebTab(string, function)"] = function(a_Simulator, a_Self, a_Title, a_Callback)
+		a_Simulator.logger:warning("Plugin uses the deprecated function cPluginLua:AddWebTab")
```
Comment (NiLSPACE) [inline APIImpl/PluginManager.lua:89]: Should we perhaps add a flag in the APIDesc if something is deprecated or not? That way we can automatically detect if a plugin is using deprecated functions.

**Review by NiLSPACE — COMMENTED** [review]
I think the `(res == true) and false or true` needs changes

File: APIImpl/PluginManager.lua (hunk 1 above)
Comment (madmaxoft) [reply APIImpl/PluginManager.lua]: Right, that's because I'm stupid and short-circuited. This needs a proper `if`.

Comment (NiLSPACE) [reply APIImpl/PluginManager.lua]: Can't you simply return `res`? If you want to be certain it's a bool you can use `not not res`

File: APIImpl/PluginManager.lua [hunk 3]
```diff
@@ -50,8 +164,9 @@ return
 	["<static> cPluginManager:GetPluginsPath()"] = function (a_Simulator)
 		return a_Simulator.options.pluginPath .. "/.."
 	end,
-	
-	["cPlugin:GetLocalFolder()"] = function(a_Simulator, a_Self)
-		return getmetatable(a_Self).simulatorInternal_pluginPath or a_Simulator:createInstance({Type = "string"})
```
Comment (NiLSPACE) [inline APIImpl/PluginManager.lua:55]: Why is this removed?

File: APIImpl/PluginManager.lua (hunk 1 above)
Comment (madmaxoft) [reply APIImpl/PluginManager.lua]: Yes, returning `not(res == true)` will do.

File: APIImpl/PluginManager.lua (hunk 3 above)
Comment (madmaxoft) [reply APIImpl/PluginManager.lua:55]: Alpha-sorted to the top of the file.

File: APIImpl/PluginManager.lua (hunk 2 above)
Comment (madmaxoft) [reply APIImpl/PluginManager.lua:89]: Yes, that would be great, but needs manpower :P

Comment (madmaxoft) [reply APIImpl/PluginManager.lua:89]: The flag should not only specify that it IS deprecated, it should also specify what to use instead. And APIDump should process this information into the documentation.

File: APIImpl/PluginManager.lua (hunk 3 above)
Comment (NiLSPACE) [reply APIImpl/PluginManager.lua:55]: Oh, I missed that, sorry.

**Comment by madmaxoft** [comment]
All fixed. I think this is ready now.

**Review by NiLSPACE — APPROVED** [review]
